Adopt more smart pointers in History.cpp

rwlbuis opened the pull request:
#### cde48409003c820470a6e7ff1f5e2c753fa956e7
<pre>
Adopt more smart pointers in History.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=273541">https://bugs.webkit.org/show_bug.cgi?id=273541</a>

Reviewed by Sihui Liu.

Adopt more smart pointers in History.cpp based on the
[alpha.webkit.UncountedLocalVarsChecker] warning.

* Source/WebCore/page/History.cpp:
(WebCore::History::length const):
(WebCore::History::scrollRestoration const):
(WebCore::History::setScrollRestoration):
(WebCore::History::go):
(WebCore::History::urlForState):

Canonical link: <a href="https://commits.webkit.org/278310@main">https://commits.webkit.org/278310@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3ecb06857b150790a761b3c2be258e0c1138531e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/50168 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29459 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2461 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53426 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/858 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52471 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35685 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/424 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/40935 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52267 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/27133 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/43174 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/22032 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24556 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/425 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8546 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/46544 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/486 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/55009 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25264 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/391 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48332 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26525 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43359 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/47345 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11002 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27388 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26257 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/page/History.cpp:
```diff
@@ -72,22 +72,22 @@ static Exception documentNotFullyActive()
         return documentNotFullyActive();
 
-    auto* historyItem = frame->history().currentItem();
+    SUPPRESS_UNCOUNTED_LOCAL auto* historyItem = frame->history().currentItem();
```
We shouldn't add `SUPPRESS_UNCOUNTED_LOCAL` I think

rwlbuis replied:
Removed.

cdumez commented on Source/WebCore/page/History.cpp:
```diff
@@ -109,19 +109,19 @@ ExceptionOr<void> History::setScrollRestoration(ScrollRestoration scrollRestorat
 SerializedScriptValue* History::stateInternal() const
 {
-    auto* frame = this->frame();
+    SUPPRESS_UNCOUNTED_LOCAL auto* frame = this->frame();
```
ditto.

rwlbuis replied:
Removed.

cdumez commented on Source/WebCore/page/History.cpp:
```diff
@@ -109,19 +109,19 @@ ExceptionOr<void> History::setScrollRestoration(ScrollRestoration scrollRestorat
     if (!frame)
         return nullptr;
-    auto* historyItem = frame->history().currentItem();
+    SUPPRESS_UNCOUNTED_LOCAL auto* historyItem = frame->history().currentItem();
```
ditto.

rwlbuis replied:
Also removed.

szewai commented on Source/WebCore/page/History.cpp:
```diff
@@ -195,7 +195,7 @@ ExceptionOr<void> History::go(Document& document, int distance)
+    RefPtr frame = this->frame();
     if (urlString.isNull())
         return frame->document()->url();
     return frame->document()->completeURL(urlString);
```
I think we need to null check document (or assert) here, and protect it before invoking `completeURL`.

rwlbuis replied:
Good point! I now use protectedDocument. For urlForState I removed reffing of the frame since the (only) caller (stateObjectAdded) already has a ref to the frame.

szewai commented on Source/WebCore/page/History.cpp:
```diff
@@ -96,11 +96,11 @@ ExceptionOr<History::ScrollRestoration> History::scrollRestoration() const
         return documentNotFullyActive();
 
-    auto* historyItem = frame->history().currentItem();
+    RefPtr historyItem = frame->history().currentItem();
```
Seems this can be simplified as:
```
if (RefPtr historyItem = frame->history().currentItem())
    historyItem->setShouldRestoreScrollPosition(scrollRestoration == ScrollRestoration::Auto);
```

rwlbuis committed:
Adopt more smart pointers in History.cpp
https://bugs.webkit.org/show_bug.cgi?id=273541

Reviewed by Sihui Liu.

Adopt more smart pointers in History.cpp based on the
[alpha.webkit.UncountedLocalVarsChecker] warning.

* Source/WebCore/page/History.cpp:
(WebCore::History::length const):
(WebCore::History::scrollRestoration const):
(WebCore::History::setScrollRestoration):
(WebCore::History::go):
(WebCore::History::urlForState):

Canonical link: https://commits.webkit.org/278310@main
